nni/mutable/symbol.py

```diff
@@ -0,0 +1,472 @@
+        return self.expr_cls(operator.{opt}, '{{}} {sym} {{}}', [other, self])"""
+
+    unary_template = """    def __{op}__(self) -> Any:
+        return self.expr_cls(operator.{op}, '{sym}{{}}', [self])"""
+
+    for op, sym in MAPPING.items():
+        if op in ['neg', 'pos', 'invert']:
+            print(unary_template.format(op=op, sym=sym) + '\n')
+        else:
+            opt = op + '_' if op in ['and', 'or'] else op
+            print(binary_template.format(op=op, opt=opt, sym=sym) + '\n')
+            if op not in ['lt', 'le', 'eq', 'ne', 'ge', 'gt']:
+                print(binary_r_template.format(op=op, opt=opt, sym=sym) + '\n')
+
+
+def symbolic_staticmethod(orig_func: Callable) -> staticmethod:
+    if orig_func.__doc__ is not None:
+        orig_func.__doc__ += """
+        Notes
+        -----
+        This function performs lazy evaluation.
+        Only the expression will be recorded when the function is called.
+        The real evaluation happens when the inner value choice has determined its final decision.
+        If no value choice is contained in the parameter list, the evaluation will be intermediate."""
+    return staticmethod(orig_func)
+
+
+def first_symbolic_object(*objects: Any) -> SymbolicExpression | None:
+    """
+    Return the first symbolic object in the given list.
+    """
+    for obj in objects:
+        if isinstance(obj, SymbolicExpression):
+            return obj
+    return None
+
+
+class SymbolicExpression:
+    """Implementation of symbolic execution.
+
+    Each instance of :class:`SymbolicExpression` is a node on the expression tree,
+    with a function and a list of children (i.e., function arguments).
+
+    The expression is designed to be compatible with native Python expressions.
+    That means, the static methods (as well as operators) can be also applied on plain Python values.
+    """
+
+    @property
+    def expr_cls(self) -> Type[SymbolicExpression]:
+        """The created expression will be using this class."""
+        return SymbolicExpression
+
+    def __init__(self, function: Callable, repr_template: str, arguments: list[Any]):
+        self.function = function
+        self.repr_template = repr_template
+        self.arguments = arguments
+
+        if not any(isinstance(arg, SymbolicExpression) for arg in self.arguments):
+            raise ValueError('To construct a symbolic expression, at least one of arguments must be a symbolic expression itself.')
+
+    def leaf_symbols(self) -> Iterable[Symbol]:
+        """
+        Return a generator of all leaf symbols.
+
+        Useful for when you want to inspect when the symbols come from.
+        No deduplication even if the symbols has duplicates.
+        """
+        for arg in self.arguments:
+            if isinstance(arg, SymbolicExpression):
+                yield from arg.leaf_symbols()
+
+    @overload
+    def evaluate(self, values: Iterable[Any]) -> Any:
```

Comment:
will the `Symbol('x') + Symbol('x')` got a 3 with input `[1, 2]`?

Reply:
Yes, but I think it doesn't really matter, because dict is more often-used now. Using a list of values, on the other hand, isn't really convenient...

Reply:
It seems that there are many tests using a list input in #5200. Is it ok?

Reply:
suggestion: try do next() in the end of the evaluate and assert getting the StopIteration

Reply:
> It seems that there are many tests using a list input in #5200. Is it ok?

Those tests are ported from a version. It always doesn't harm to have more tests.

> suggestion: try do next() in the end of the evaluate and assert getting the StopIteration

Might need to rewrite the whole `evaluate` to make this possible. Not sure whether worth the effort.